src/library/scala/collection/immutable/NumericRange.scala

```diff
@@ -340,10 +340,18 @@ object NumericRange {
           // Jump in three pieces:
           //   * start to -1 or 1, whichever is closer (waypointA)
           //   * one step, which will take us at least to 0 (ends at waypointB)
+          //   * (except with really small numbers)
           //   * there to the end
           val negone = num.fromInt(-1)
           val startlim  = if (posStep) negone else one
-          val startdiff = num.minus(startlim, start)
+          //Use start value if the start value is closer to zero than startlim
+          //   * e.g. .5 is closer to zero than 1 and -.5 is closer to zero than -1
+          val startdiff = {
+            if ((posStep && num.lt(startlim, start)) || (!posStep && num.gt(startlim, start)))
+              start
```

Comment:
I don't know this code, but what surprises me here is that `startdiff` can be negative in the first case, if `posStep && num.lt(startlim, start)`.

Reading on, `waypointA` can be smaller than `start`. Is that intended?

Maybe it just doesn't matter, but I'd like to double-check.

Reply:
So `startdiff` is negative when crossing zero from negative to positive and the start is greater than -1 (between -1 and 0).
For example, in the original code, if  `start == -.75`, then `startdiff == -.25`.

`waypointA` can be smaller than start because it trips [num.lt(waypointB, end)]https://github.com/scala/scala/pull/9155/files#diff-82fe528d34183eb1be9c0c90e48cb447R359).  

[This test case](https://github.com/scala/scala/pull/9155/files#diff-b1c2caabda073bf0b48349132118c517R119), where range is `Range.BigDecimal(BigDecimal(-2E-34), BigDecimal(1E-64), BigDecimal(1E-34))` is an example of when `waypointA` would be less than `start` and how it works.

Apologies for the subpar explanation; if needed, I can work back through the changes and better document the what/why.

Reply:
OK, thank you for confirming. I looked a bit more at the code with a number of examples and I see now how your fix works. Together with the new and existing tests, LGTM.